File: Programmers/Lev2/Stock_Cost/Stock_Cost.cpp
#include <string>
#include <vector>
#include <iostream>
#include<stack>

using namespace std;


vector<int> solution(vector<int> prices) {
    int length = prices.size();

    vector<int> answer(length, 0);

    stack<pair<int, int>> my_stack;

    for (int current_index = 0; current_index < length; current_index++) {
        if (my_stack.size() < 1) {
            my_stack.push({ prices[current_index], current_index });
        }
        else {
            int t_last_index = my_stack.size()-1;
            pair<int, int> t_element = my_stack.top();
            
            while ((my_stack.size() > 0)& (t_element.first > prices[current_index])) {
                answer[t_element.second] = current_index - t_element.second;
                my_stack.pop();
                if (my_stack.size() > 0) {
                    t_element = my_stack.top();
                }
            }

            my_stack.push({ prices[current_index], current_index });
        }
    }

    while (my_stack.size() > 0) {
        pair<int, int> t_ele = my_stack.top();
        answer[t_ele.second] = (length - 1) - t_ele.second;
        my_stack.pop();
    }

    return answer;
}

int main() {
    vector<int> inputs = {5,5,3,3,4,3,1};

    vector<int> result = solution(inputs);

    for (int i = 0; i < result.size(); i++) {
        cout << result[i] << " ";
    }
   
}

File: Programmers/Lev2/Stock_Cost/Stock_Cost.py
def solution(prices):
    answer = []

    # answer 0으로 초기화
    answer = [0 for i in range(len(prices))]

    my_stack = []

    for current_index in range(0, len(prices)):
        if len(my_stack) < 1:
            my_stack.append([prices[current_index], current_index])
        else:
            t_element = my_stack[-1]

            while (len(my_stack) > 0) and (t_element[0] > prices[current_index]):
                answer[t_element[1]] = current_index - t_element[1]
                my_stack.pop()
                if len(my_stack)>0:
                    t_element = my_stack[-1]

            my_stack.append([prices[current_index], current_index])


    while len(my_stack) > 0:
        t_ele = my_stack.pop()
        t_index = t_ele[1]
        answer[t_index] = current_index - t_index

    return answer

if __name__ == '__main__':
    input = [[1,2,3,2,3], [3,3,3,3,3,3], [5,5,3,3,4,3,1], [5,2,6,3]]
    expect_result = [[4,3,1,1,0], [5,4,3,2,1,0], [2,1,4,3,1,1,0], [1, 2, 1, 0]]

    for i in range(0, len(input)):
        print(solution(input[i]), expect_result[i])
